src/main/kotlin/uk/gov/gdx/datashare/services/SupplierEventProcessor.kt

```diff
@@ -0,0 +1,91 @@
+import org.springframework.transaction.annotation.Transactional
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchRequest
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchRequestEntry
+import uk.gov.gdx.datashare.config.SupplierSubscriptionNotFoundException
+import uk.gov.gdx.datashare.queue.AwsQueue
+import uk.gov.gdx.datashare.queue.AwsQueueService
+import uk.gov.gdx.datashare.repositories.*
+
+@Service
+@XRayEnabled
+class SupplierEventProcessor(
+  private val objectMapper: ObjectMapper,
+  private val supplierSubscriptionRepository: SupplierSubscriptionRepository,
+  private val acquirerSubscriptionRepository: AcquirerSubscriptionRepository,
+  private val supplierEventRepository: SupplierEventRepository,
+  private val awsQueueService: AwsQueueService,
+
+) {
+  private val acquirerEventQueue by lazy { awsQueueService.findByQueueId("acquirerevent") as AwsQueue }
+  private val acquirerEventSqsClient by lazy { acquirerEventQueue.sqsClient }
+  private val acquirerEventQueueUrl by lazy { acquirerEventQueue.queueUrl }
+
+  companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  @JmsListener(destination = "supplierevent", containerFactory = "awsQueueContainerFactoryProxy")
+  @Transactional
+  fun onSupplierEvent(message: String) {
+    val supplierEvent = buildSupplierEvent(message)
+    val supplierSubscription = fetchSupplierSubscription(supplierEvent)
+    supplierEventRepository.save(supplierEvent)
+    val acquirerEvents = buildAcquirerEvents(supplierSubscription, supplierEvent)
+    enqueueAcquirerEvents(acquirerEvents)
+  }
+
+  private fun buildSupplierEvent(message: String): SupplierEvent {
+    return objectMapper.readValue(message, SupplierEvent::class.java)
+  }
+
+  private fun fetchSupplierSubscription(supplierEvent: SupplierEvent) =
+    (
+      supplierSubscriptionRepository.findByIdOrNull(supplierEvent.supplierSubscriptionId)
+        ?: throw SupplierSubscriptionNotFoundException(supplierEvent.supplierSubscriptionId)
+      )
+
+  private fun buildAcquirerEvents(
+    supplierSubscription: SupplierSubscription,
+    supplierEvent: SupplierEvent,
+  ): List<AcquirerEvent> {
+    val acquirerSubscriptions = acquirerSubscriptionRepository.findAllByEventType(supplierSubscription.eventType)
+    val acquirerEvents = acquirerSubscriptions.map { acquirerSubscription ->
+      AcquirerEvent(
+        supplierEventId = supplierEvent.id,
+        dataId = supplierEvent.dataId,
+        eventTime = supplierEvent.eventTime,
+        acquirerSubscriptionId = acquirerSubscription.id,
+      )
+    }
+    return acquirerEvents
+  }
+
+  private fun enqueueAcquirerEvents(acquirerEvents: List<AcquirerEvent>) {
+    val acquirerEventMessages = acquirerEvents
+      .map {
+        SendMessageBatchRequestEntry.builder()
+          .id(it.id.toString())
+          .messageBody(it.toJson())
+          .build()
+      }
+
+    acquirerEventMessages.chunked(10) {
```

Comment:
should this be configurable?

Reply:
No, this is dictated by AWS. We want to minimise the number of requests for billing reasons, and they accept up to 10 messages per batch

Reply:
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SendMessageBatch.html

Reply:
Otherwise we'd send them all in one batch and make it a transaction

Reply:
Sounds good to me, was unsure where the 10 came from :)